README.md
# Fork Sync
[![Build](https://github.com/tg908/fork-sync/workflows/PR%20Checks/badge.svg)](https://github.com/tg908/fork-sync/actions?workflow=PR%20Checks)
![Version](https://img.shields.io/github/v/release/tg908/fork-sync?style=flat-square)

Github action to sync your Forks.
This action uses octokit and the GitHub API to automatically create and merge a pull request with the head defined by `head` into the base defined by `base`. The head branch owner is defined by `owner`. If you create a PR in the same repository you can omit the `owner` parameter.

# Example Workflow

```yml
name: Sync Fork

on:
  schedule:
    - cron: '*/30 * * * *' # every 30 minutes
  workflow_dispatch: # on button click

jobs:
  sync:

    runs-on: ubuntu-latest

    steps:
      - uses: tgymnich/fork-sync@v1.4
        with:
          owner: llvm
          base: master
          head: master
```

## Auto approve

If you use a workflow which does not allow to merge pull requests without a review 
("Require pull request reviews before merging" in your [repo settings](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/configuring-protected-branches))
you can set `auto_approve` to `true`. In that case you'll have to provide a [personal access token](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token)
for a user which is allowed to review the pull requests changes. Make sure the token has at least
`public_repo` permissions and store the token inside of the [repository secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository).

An example workflow would then look like this:

```yml
name: Sync Fork

on:
  schedule:
    - cron: '*/30 * * * *' # every 30 minutes
  workflow_dispatch: # on button click

jobs:
  sync:

    runs-on: ubuntu-latest

    steps:
      - uses: tgymnich/fork-sync@v1.4
        with:
          token: ${{ secrets.PERSONAL_TOKEN }}
          owner: llvm
          base: master
          head: master
```

# Parameters

|  name           |   Optional  |   Default              |   description                                        |
|---              |---          |---                     |---                                                   |
|   owner         | ✅          | $current_repo_owner    |   Owner of the forked repository                     |
|   token         | ✅          | ${{ github.token }}    |   Token  to access the Github API                    |
|   head          | ✅          | master                 |   Head branch                                        |
|   base          | ✅          | master                 |   Base branch                                        |
|   merge_method  | ✅          | merge                  |   merge, rebase or squash                            |
|   pr_title      | ✅          | Fork Sync              |   Title of the created pull request                  |
|   pr_message    | ✅          |                        |   Message of the created pull request                |
|   ignore_fail   | ✅          |                        |   Ignore Exceptions                                  |

⚠️ $current_repo_owner is your own username!

⚠️ Only provide the branch name for `head` and `base`. `user:branch` will not work! 

⚠️ * if `auto_approve` is set to `true` you must provide a personal access token in `token` the default github token won't work! 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## Types of changes

* Added - for new features.
* Changed - for changes in existing functionality.
* Deprecated - for soon-to-be removed features.
* Removed - for now removed features.
* Fixed - for any bug fixes.
* Security - in case of vulnerabilities.

## [Unreleased]

### Added

* Default workflows for Tartufo and super-linter

### Changed

* Updated with super-linter formatting/linting recommendations

### Deprecated

### Removed

* Removed all remnants of codecov

### Fixed


## [1.0.0] - Feb. 1, 2021

* Base release

## Attribution

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

[unreleased]: https://github.com/gdcorp-action-public-forks/fork-sync/compare/v1.0.0...HEAD
[comment]: <> ([1.1.0]: https://github.com/gdcorp-action-public-forks/fork-sync/compare/v1.0.0...v1.1.0)
[1.0.0]: https://github.com/gdcorp-action-public-forks/fork-sync/releases/tag/v1.0.0


src/main.ts
import * as core from '@actions/core';
const Github = require('@actions/github');
const { Octokit } = require("@octokit/rest");
const { retry } = require("@octokit/plugin-retry");
const token = core.getInput('token', { required: true });
const context = Github.context;
const MyOctokit = Octokit.plugin(retry);

async function run() {
  const owner = core.getInput('owner', { required: false }) || context.repo.owner;
  const base = core.getInput('base', { required: false });
  const head = core.getInput('head', { required: false });
  const mergeMethod = core.getInput('merge_method', { required: false });
  const prTitle = core.getInput('pr_title', { required: false });
  const prMessage = core.getInput('pr_message', { required: false });
  const ignoreFail = core.getInput('ignore_fail', { required: false });
  const autoApprove = core.getInput('auto_approve', { required: false });
  const retries = parseInt(core.getInput('retries', { required: false })) || 4;
  const retryAfter = parseInt(core.getInput('retry_after', { required: false })) || 60;
  
  const octokit = new MyOctokit({
    auth: token,
    request: {
      retries,
      retryAfter,
    },
  });

  try {
    let pr = await octokit.pulls.create({ owner: context.repo.owner, repo: context.repo.repo, title: prTitle, head: owner + ':' + head, base: base, body: prMessage, merge_method: mergeMethod, maintainer_can_modify: false });
    await delay(20);
    if (autoApprove) {
        await octokit.pulls.createReview({ owner: context.repo.owner, repo: context.repo.repo, pull_number: pr.data.number, event: "COMMENT", body: "Auto approved" });
        await octokit.pulls.createReview({ owner: context.repo.owner, repo: context.repo.repo, pull_number: pr.data.number, event: "APPROVE" });
    }
    await octokit.pulls.merge({ owner: context.repo.owner, repo: context.repo.repo, pull_number: pr.data.number });
  } catch (error) {
    if (error.request.request.retryCount) {
      console.log(
        `request failed after ${error.request.request.retryCount} retries with a delay of ${error.request.request.retryAfter}`
      );
    }
    if (!!error.errors && !!error.errors[0] && !!error.errors[0].message && error.errors[0].message.startsWith('No commits between')) {
      console.log('No commits between ' + context.repo.owner + ':' + base + ' and ' + owner + ':' + head);
    } else {
      if (!ignoreFail) {
        core.setFailed(`Failed to create or merge pull request: ${error ?? "[n/a]"}`);
      }
    }
  }
}

function delay(s: number) {
  return new Promise( resolve => setTimeout(resolve, s * 1000) );
}

run();


super-linter.sh
#!/bin/bash
docker pull github/super-linter:latest

# shellcheck disable=SC2034
PATH_TO_CODEBASE="$(pwd)"
VALIDATE_ALL_CODEBASE="false"
#VALIDATE_ANSIBLE="false"
#VALIDATE_ARM="false"
VALIDATE_BASH="false"
VALIDATE_BASH_EXEC="false"
#VALIDATE_CLOJURE="false"
VALIDATE_CLOUDFORMATION="false"
#VALIDATE_COFFEE="false"
VALIDATE_CSHARP="false"
#VALIDATE_CSS="false"
#VALIDATE_DART="false"
#VALIDATE_DOCKERFILE="false"
VALIDATE_DOCKERFILE_HADOLINT="false"
VALIDATE_EDITORCONFIG="false"
#VALIDATE_ENV="false"
#VALIDATE_GHERKIN="false"
#VALIDATE_GO="false"
#VALIDATE_GROOVY="false"
#VALIDATE_HTML="false"
VALIDATE_JAVA="false"
#VALIDATE_JAVASCRIPT_ES="false"
#VALIDATE_JAVASCRIPT_STANDARD="false"
VALIDATE_JSCPD="false"
VALIDATE_JSON="false"
#VALIDATE_JSX="false"
#VALIDATE_KOTLIN="false"
#VALIDATE_KUBERNETES_KUBEVAL="false"
#VALIDATE_LATEX="false"
#VALIDATE_LUA="false"
#VALIDATE_MARKDOWN="false"
#VALIDATE_OPENAPI="false"
#VALIDATE_PERL="false"
#VALIDATE_PHP="false"
#VALIDATE_PHP_BUILTIN="false"
#VALIDATE_PHP_PHPCS="false"
#VALIDATE_PHP_PHPSTAN="false"
#VALIDATE_PHP_PSALM="false"
#VALIDATE_PROTOBUF="false"
VALIDATE_PYTHON="false"
#VALIDATE_PYTHON_BLACK="false"
VALIDATE_PYTHON_FLAKE8="false"
VALIDATE_PYTHON_ISORT="false"
VALIDATE_PYTHON_MYPY="false"
VALIDATE_PYTHON_PYLINT="false"
VALIDATE_POWERSHELL="false"
#VALIDATE_R="false"
#VALIDATE_RAKU="false"
#VALIDATE_RUBY="false"
#VALIDATE_RUST_2015="false"
#VALIDATE_RUST_2018="false"
#VALIDATE_RUST_CLIPPY="false"
VALIDATE_SHELL_SHFMT="false"
#VALIDATE_SNAKEMAKE_LINT="false"
#VALIDATE_SNAKEMAKE_SNAKEFMT="false"
#VALIDATE_STATES="false"
#VALIDATE_SQL="false"
#VALIDATE_TEKTON="false"
#VALIDATE_TERRAFORM="false"
#VALIDATE_TERRAFORM_TERRASCAN="false"
#VALIDATE_TERRAGRUNT="false"
#VALIDATE_TSX="false"
#VALIDATE_TYPESCRIPT_ES="false"
#VALIDATE_TYPESCRIPT_STANDARD="false"
VALIDATE_XML="false"
VALIDATE_YAML="false"

docker run -e RUN_LOCAL=true -e VALIDATE_JAVA=${VALIDATE_JAVA} \
    -e VALIDATE_ALL_CODEBASE=${VALIDATE_ALL_CODEBASE} \
    -e VALIDATE_CLOUDFORMATION=${VALIDATE_CLOUDFORMATION} \
    -e VALIDATE_CSHARP=${VALIDATE_CSHARP} \
    -e VALIDATE_EDITORCONFIG=${VALIDATE_EDITORCONFIG} \
    -e VALIDATE_JAVA=${VALIDATE_JAVA} \
    -e VALIDATE_JSCPD=${VALIDATE_JSCPD} \
    -e VALIDATE_JSON=${VALIDATE_JSON} \
    -e VALIDATE_PYTHON=${VALIDATE_PYTHON} \
    -e VALIDATE_PYTHON_FLAKE8=${VALIDATE_PYTHON_FLAKE8} \
    -e VALIDATE_PYTHON_ISORT=${VALIDATE_PYTHON_ISORT} \
    -e VALIDATE_PYTHON_MYPY=${VALIDATE_PYTHON_MYPY} \
    -e VALIDATE_PYTHON_PYLINT=${VALIDATE_PYTHON_PYLINT} \
    -e VALIDATE_BASH=${VALIDATE_BASH} \
    -e VALIDATE_BASH_EXEC=${VALIDATE_BASH_EXEC} \
    -e VALIDATE_SHELL_SHFMT=${VALIDATE_SHELL_SHFMT} \
    -e VALIDATE_XML=${VALIDATE_XML} \
    -e VALIDATE_YAML=${VALIDATE_YAML} \
    -e VALIDATE_POWERSHELL=${VALIDATE_POWERSHELL} \
    -e VALIDATE_DOCKERFILE_HADOLINT=${VALIDATE_DOCKERFILE_HADOLINT} \
    -e IGNORE_GITIGNORED_FILES=true \
    -v "${PATH_TO_CODEBASE}":/tmp/lint github/super-linter
#    -e VALIDATE_PYTHON_BLACK=${VALIDATE_PYTHON_BLACK} \
#    -e VALIDATE_MARKDOWN=${VALIDATE_MARKDOWN} \
#    -e VALIDATE_ENV=${VALIDATE_ENV} \
#    -e VALIDATE_R=${VALIDATE_R} \
#    -e VALIDATE_RAKU=${VALIDATE_RAKU} \
#    -e VALIDATE_RUBY=${VALIDATE_RUBY} \
#    -e VALIDATE_RUST_2015=${VALIDATE_RUST_2015} \
#    -e VALIDATE_RUST_2018=${VALIDATE_RUST_2018} \
#    -e VALIDATE_RUST_CLIPPY=${VALIDATE_RUST_CLIPPY} \
#    -e VALIDATE_SNAKEMAKE_LINT=${VALIDATE_SNAKEMAKE_LINT} \
#    -e VALIDATE_SNAKEMAKE_SNAKEFMT=${VALIDATE_SNAKEMAKE_SNAKEFMT} \
#    -e VALIDATE_STATES=${VALIDATE_STATES} \
#    -e VALIDATE_SQL=${VALIDATE_SQL} \
#    -e VALIDATE_TEKTON=${VALIDATE_TEKTON} \
#    -e VALIDATE_TERRAFORM=${VALIDATE_TERRAFORM} \
#    -e VALIDATE_TERRAFORM_TERRASCAN=${VALIDATE_TERRAFORM_TERRASCAN} \
#    -e VALIDATE_TERRAGRUNT=${VALIDATE_TERRAGRUNT} \
#    -e VALIDATE_TSX=${VALIDATE_TSX} \
#    -e VALIDATE_TYPESCRIPT_ES=${VALIDATE_TYPESCRIPT_ES} \
#    -e VALIDATE_TYPESCRIPT_STANDARD=${VALIDATE_TYPESCRIPT_STANDARD} \
#    -e VALIDATE_LATEX=${VALIDATE_LATEX} \
#    -e VALIDATE_LUA=${VALIDATE_LUA} \
#    -e VALIDATE_OPENAPI=${VALIDATE_OPENAPI} \
#    -e VALIDATE_PERL=${VALIDATE_PERL} \
#    -e VALIDATE_PHP=${VALIDATE_PHP} \
#    -e VALIDATE_PHP_BUILTIN=${VALIDATE_PHP_BUILTIN} \
#    -e VALIDATE_PHP_PHPCS=${VALIDATE_PHP_PHPCS} \
#    -e VALIDATE_PHP_PHPSTAN=${VALIDATE_PHP_PHPSTAN} \
#    -e VALIDATE_PHP_PSALM=${VALIDATE_PHP_PSALM} \
#    -e VALIDATE_PROTOBUF=${VALIDATE_PROTOBUF} \
#    -e VALIDATE_JSX=${VALIDATE_JSX} \
#    -e VALIDATE_KOTLIN=${VALIDATE_KOTLIN} \
#    -e VALIDATE_JAVASCRIPT_ES=${VALIDATE_JAVASCRIPT_ES} \
#    -e VALIDATE_JAVASCRIPT_STANDARD=${VALIDATE_JAVASCRIPT_STANDARD} \
#    -e VALIDATE_GHERKIN=${VALIDATE_GHERKIN} \
#    -e VALIDATE_GO=${VALIDATE_GO} \
#    -e VALIDATE_GROOVY=${VALIDATE_GROOVY} \
#    -e VALIDATE_HTML=${VALIDATE_HTML} \
#    -e VALIDATE_COFFEE=${VALIDATE_COFFEE} \
#    -e VALIDATE_CSS=${VALIDATE_CSS} \
#    -e VALIDATE_DART=${VALIDATE_DART} \
#    -e VALIDATE_ANSIBLE=${VALIDATE_ANSIBLE} \
#    -e VALIDATE_ARM=${VALIDATE_ARM} \
#    -e VALIDATE_CLOJURE=${VALIDATE_CLOJURE} \
#    -e VALIDATE_DOCKERFILE=${VALIDATE_DOCKERFILE} \


lib/main.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const core = __importStar(require("@actions/core"));
const Github = require('@actions/github');
const { Octokit } = require("@octokit/rest");
const { retry } = require("@octokit/plugin-retry");
const token = core.getInput('token', { required: true });
const context = Github.context;
const MyOctokit = Octokit.plugin(retry);
function run() {
    return __awaiter(this, void 0, void 0, function* () {
        const owner = core.getInput('owner', { required: false }) || context.repo.owner;
        const base = core.getInput('base', { required: false });
        const head = core.getInput('head', { required: false });
        const mergeMethod = core.getInput('merge_method', { required: false });
        const prTitle = core.getInput('pr_title', { required: false });
        const prMessage = core.getInput('pr_message', { required: false });
        const ignoreFail = core.getInput('ignore_fail', { required: false });
        const autoApprove = core.getInput('auto_approve', { required: false });
        const retries = parseInt(core.getInput('retries', { required: false })) || 4;
        const retryAfter = parseInt(core.getInput('retry_after', { required: false })) || 60;
        const octokit = new MyOctokit({
            auth: token,
            request: {
                retries,
                retryAfter,
            },
        });
        try {
            let pr = yield octokit.pulls.create({ owner: context.repo.owner, repo: context.repo.repo, title: prTitle, head: owner + ':' + head, base: base, body: prMessage, merge_method: mergeMethod, maintainer_can_modify: false });
            yield delay(20);
            if (autoApprove) {
                yield octokit.pulls.createReview({ owner: context.repo.owner, repo: context.repo.repo, pull_number: pr.data.number, event: "COMMENT", body: "Auto approved" });
                yield octokit.pulls.createReview({ owner: context.repo.owner, repo: context.repo.repo, pull_number: pr.data.number, event: "APPROVE" });
            }
            yield octokit.pulls.merge({ owner: context.repo.owner, repo: context.repo.repo, pull_number: pr.data.number });
        }
        catch (error) {
            if (error.request.request.retryCount) {
                console.log(`request failed after ${error.request.request.retryCount} retries with a delay of ${error.request.request.retryAfter}`);
            }
            if (!!error.errors && !!error.errors[0] && !!error.errors[0].message && error.errors[0].message.startsWith('No commits between')) {
                console.log('No commits between ' + context.repo.owner + ':' + base + ' and ' + owner + ':' + head);
            }
            else {
                if (!ignoreFail) {
                    core.setFailed(`Failed to create or merge pull request: ${error !== null && error !== void 0 ? error : "[n/a]"}`);
                }
            }
        }
    });
}
function delay(s) {
    return new Promise(resolve => setTimeout(resolve, s * 1000));
}
run();
